dynamiqs/solvers/utils/batching.py

```diff
@@ -5,16 +5,36 @@
 
 def batch_H(H: TDTensor) -> TDTensor:
     # handle H batching
-    # x: (b_H?, n, n) ->  (b_H, 1, n, n)
+    # x: (b_H?, n, n) ->  (b_H, 1, 1, n, n)
     if H.dim() == 2:  # (n, n)
         H = H.unsqueeze(0)  # (1, n, n)
-    return H.unsqueeze(1)
+    H = H.unsqueeze(1)
+    H = H.unsqueeze(1)
+    return H
 
 
-def batch_y0(y0: Tensor, H: TDTensor) -> Tensor:
+def batch_jump_ops(jump_ops: Tensor) -> Tensor:
+    # handle jump_ops batching
+    # jump_ops: (n_jump_ops, b_jump_ops?, m, n) ->  (n_jump_ops, 1, b_jump_ops, 1, n, n)
```

Comment:
I think we should have `jump_ops` of shape `(b_jump_ops?, n_jump_ops, m, n)` instead of `(n_jump_ops, b_jump_ops?, m, n)`. For 2 reasons:

1. We keep the optional dimension outward so the last dimensions are always consistent (batched or not).
2. It's more consistent with other batchable objects, with the batched dimension first.

In case it's more annoying in the backend, we can just swap the two dimensions together right here.

Reply:
We decided to ease the most common path here, which is not batched jump operators. The easiest way of passing mutiple jump operators is through a list (stacking tensors requires more code and a little gymnastic). So the best way of writing in the most common case remains `dq.mesolve(H, [L1, L2], ...)` IMO.

If we chose to put the batching direction outside we'd have to require
- either the jump operators to always be reshaped with `(b_jump_ops?, n_jump_ops, m, n)`, which is annoying
- or to have a different syntax for batched and not batched operators
- or to provide with a list of list of tensors if we want to keep not batched operators, which also sounds like a bad idea

Also I'm not sure we have other batchable and stackable (ie there are mulitple) arguments. In this case it is more consistant with other batched object to NOT put the `n_jump_ops` outside this it's not a batching dimension but a "stacking" one. Otherwise you'd have
- `y0`: `(b_H, b_L, b_y0, m, n)`
- `H`: `(b_H, 1, 1, m, n)`
- `L`: `(b_H, b_L, n_L, 1, m, n)`
which is an additional complication

Reply:
Indeed I didn't think of the fact that we pass jump operators as a list. In this case I'm good with keeping this order.

One question: how do we handle the fact that we may want multiple to pass multiple jump operators that are batched? Or alternatively, when we want to batch over a single jump operator and not all?

For instance, I want to simulate a dissipative cat qubit with `L2 = sqrt(kappa2) * (a^2 - alpha^2)` (with `kappa2` fixed), but also batch over `L1 = sqrt(kappa1) a` (with `kappa1` batched). Should I pass `L2` in a batch with a similar shape as `L1`? Should only `L1` be batched? 

Alternatively, what if I want to batch over `kappa2` **and** `kappa1`? What is the accepted input shape?

I don't think there's a particular preferred choice, but we need to make one. Could you clarify this in the PR? Maybe add checks that the input is of the required shape format?

Reply:
This is a good question. As of now all dissipators are expected to have the same batching. That being said, it's still possible to code your example with something like this (might not work out of the box but you'll get the idea):
```python
kappa_2 = 10
kappa_1 = torch.linspace(0, 1, 11)

L2  = kappa_2 * (a @ a).repeat(len(kappa_1), 1, 1)
L1  = kappa_1 a.repeat(len(kappa_1), 1, 1)

qt.mesolve(H, [L2, L1], ...)
```

It's not as straight forward as specifying `L2 = kappa_2 * a @ a` but it may be a good thing, because otherwise there is a topic on how you handle dissipators with different batch sizes etc...

If you want to batch over $\kappa_2$ and $\kappa_1$, probably something like this would work:
```
kappa_2 = torch.linspace(1, 10, 10)
kappa_1 = torch.linspace(0, 1, 11)

kappa_2, kappa_1 = torch.meshgrid(kappa_2, kappa_1)
kappa_2 = kappa_2.ravel()
kappa_1 = kappa_1.ravel()

L2  = kappa_2 * (a @ a).repeat(len(kappa_1), 1, 1)
L1  = kappa_1 a.repeat(len(kappa_1), 1, 1)

qt.mesolve(H, [L2, L1], ...)
```

It's not as easy as if we direclty made the product in `mesolve` but doing so would disqualify all cases where we want to have hand made batching (like `(x1, y1), (x1, y2), (x2, y1)` but not `(x2, y2)`). To simplify a bit I would add a helper method that would make the carthesian product of several batched tensors.

I'll add the check for same dimensionality in this PR but let's keep API changes (if we want them) for another one

Reply:
Sorry, don't have the time to do a full review right now (but pls wait for me for merging @abocquet), but on this specific point (multi-batched Hamiltonian, initial state on jump operators), I think we should:
- keep the current `sesolve`, `mesolve`, `smesolve` APIs, it's going to be unreadable with arbitrarily mulyi-batched `H`, `rho0` and `L`
- one day, if needed, provide simple utility functions to convert a multi-batched object to a single-batched object and the reverse (because internally we'll always handle a large Hamiltonian), but it's a bit of an edge case

So this feature is not really missing from the API, we just miss a bit of sugar. What IS missing is the ability to run non-product simulations of batched parameters (i.e. we run today all H of a batched Hamiltonian with all rho0 of a batched initial states, you can't _choose_ specific configurations). I think it's quite an edge case though, I don't think we want to complexify the API or put effort on this.

Reply:
100% agreed, I think we are on the same page

Conclusion : 
- until further notice the niche examples Ronan gave are to be done with the code examples in my n-1 message (give or take)
- multi-batching is something that we should think of someday but until there is a usecase/something else to do, let's put it in a corner of our heads and talk about it on the occasion

Reply:
If I understood correctly, we only accept that all jump operators have the same batching dimension for now. Can you just add a check that this is indeed the case somewhere, and raise an error if not?

Reply:
Actually, to simplify usability, I would propose that the following two ways of batching are allowed:

1. All jump operators have the same batching dimension `b_L`.
2. A single jump operator has a batching dimension `b_L`, and all others have no batching dimensions (`ndim=2`).

In case 2, we simply batch all other jump operators internally with the same shape. All other batching shapes should be disallowed. WDYT?

Reply:
I'm fine with either propositions (either force all batched with same dim, or accept both all batched with same dim and a single one batched). Let's not document this new feature in mesolve documentation or in the batching tutorial for now, it needs a bit more thinking and experimentations in the backend before going public.

Reply:
Addressed in https://github.com/dynamiqs/dynamiqs/pull/325/commits/f20ab9860564027a287fbd8aea8dcc80f46302bb